### PR TITLE
GH-394: Adjust module wrapping to contain output folder value based on the manifest

### LINKED ISSUE #394: Adjust module wrapping to contain output folder value based on the manifest
Rising popularity of the JS transpilers (babel, type script) causes new pattern to emerge in npm libraries: they contain both source and compiled output.

Changing module wrapping in the compiler would comply with this pattern. This would also simplify our npm exports of the N4JS projects, where currently we need to move compiled code to the root folder.

`Output` field from manifest is not the only extra segment in the compiled code location.
1. What should we do about `compiler id` that comes from multiple transpilers existing for N4JS?
2. What should we do about `project id` (and version)?

AC:
* adjust module wrapping

1. remove project name from output segments
2. remove compiler id (at least for now, later multiple compilers might bring segment back)
3. attach output folder to the module wrapping (i.e. emitted SystemJS imports)

* adjust `NODE_PATH` in runners and testers
* adjust NPM import/export
* adjust NFAR import/export

### PR DESCRIPTION
**Pull request by yoosiba**
resolves #394

Changes:
* files are into project output path, without adding `projectID` and `compilerID` segments
* imports are resolved from project root (and don't include dropped segments)
* paths for runners and testers are adjusted accordingly
* bootstrapping runtime environments are adjusted accordingly
* paths for npm export are adjusted accordingly

### CODE REVIEW

File: tests/org.eclipse.n4js.spec.examples.xpect.tests/testdata/my/Ex65_Object_Literal.n4js.xt [hunk 1]
```diff
@@ -57,7 +57,7 @@ stderr:
 			}
 		};
 	});
-})(typeof module !== 'undefined' && module.exports ? require('n4js-node/index').System(require, module) : System);
+})(typeof module !== 'undefined' && module.exports ? require('n4js-node/src-gen/index').System(require, module) : System);
```
Comment (mor-n4) [inline tests/org.eclipse.n4js.spec.examples.xpect.tests/testdata/my/Ex65_Object_Literal.n4js.xt]: You changed several spec example tests. But it seems you did not update the examples in the spec accordingly?

File: plugins/org.eclipse.n4js.tester/src/org/eclipse/n4js/tester/TestDiscoveryHelper.java [hunk 2]
```diff
@@ -94,7 +95,7 @@
 	@Inject
 	private IN4JSCore n4jsCore;
 	@Inject
-	private QualifiedNameComputer qualifiedNameComputer;
+	private ResourceNameComputer qualifiedNameComputer;
```
Comment (mor-n4) [inline plugins/org.eclipse.n4js.tester/src/org/eclipse/n4js/tester/TestDiscoveryHelper.java]: please update the field name accordingly (please also check in other places)

File: plugins/org.eclipse.n4js.transpiler/src/org/eclipse/n4js/transpiler/AbstractTranspiler.java [hunk 3]
```diff
@@ -57,7 +57,7 @@
 	private ScopeManager scopeManager;
 
 	@Inject
-	private QualifiedNameComputer qnComputer;
+	private ResourceNameComputer qnComputer;
```
Comment (mor-n4) [inline plugins/org.eclipse.n4js.transpiler/src/org/eclipse/n4js/transpiler/AbstractTranspiler.java]: adjust field name

File: plugins/org.eclipse.n4js.transpiler.es/src/org/eclipse/n4js/transpiler/es/transform/ExpressionTransformation.xtend [hunk 4]
```diff
@@ -49,7 +49,7 @@ import static extension org.eclipse.n4js.typesystem.RuleEnvironmentExtensions.*
 @ExcludesBefore(AsyncAwaitTransformation)
 class ExpressionTransformation extends Transformation {
 
-	@Inject private QualifiedNameComputer qualifiedNameComputer;
+	@Inject private ResourceNameComputer qualifiedNameComputer;
```
Comment (mor-n4) [inline plugins/org.eclipse.n4js.transpiler.es/src/org/eclipse/n4js/transpiler/es/transform/ExpressionTransformation.xtend]: field name

File: tests/org.eclipse.n4js.lang.tests/src/org/eclipse/n4js/naming/QualifiedNameComputerTest.java [hunk 5]
```diff
@@ -43,7 +42,7 @@
 	private Type type;
 
 	@Inject
-	QualifiedNameComputer qnc;
+	ResourceNameComputer qnc;
```
Comment (mor-n4) [inline tests/org.eclipse.n4js.lang.tests/src/org/eclipse/n4js/naming/QualifiedNameComputerTest.java]: field name

ALSO: this class should be renamed (`QualifiedNameComputerTest` is obsolete after removing the `QualifiedNameComputer`)

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/projectModel/FindArtifactHelper.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+import org.eclipse.xtext.naming.IQualifiedNameConverter;
+import org.eclipse.xtext.naming.QualifiedName;
+
+import com.google.common.base.Optional;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+
+/**
+ * Helper for finding artifacts in the projects. Delegates to the
+ * {@link IN4JSSourceContainer#findArtifact(QualifiedName, Optional)} but exposes more convinient API.
```
Comment (mor-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/projectModel/FindArtifactHelper.java]: typo in 2nd to last word

File: plugins/org.eclipse.n4js.npmexporter/src/org/eclipse/n4js/npmexporter/NpmExporter.xtend [hunk 7]
```diff
@@ -104,10 +101,7 @@ class NpmExporter {
 		}
 
 		// copy files from original output folder to targetFolder
-//		val Path project_loc = project.locationPath;
-		val String project_out = project.outputPath;
-		// TODO improve generator-access, here ES-transpiler is directly referenced
```
[the remark this answers is not in the document]
Comment (yoosiba) [reply plugins/org.eclipse.n4js.npmexporter/src/org/eclipse/n4js/npmexporter/NpmExporter.xtend]: Yes, so mentioned ES transpiler is not directly accessed, and there is only one output folder. So comment is not relevant. When we add new generators and each will have its own output folder, then we need to change N4MF which will cause the project model to be adjust, thus accessing `project.outputPath` will fail (as in the file will not compile).
IMHO TODO in question is obsolete.

File: plugins/org.eclipse.n4js.runner.nodejs/src/org/eclipse/n4js/runner/nodejs/NodeEngineCommandBuilder.java [hunk 8]
```diff
@@ -73,71 +76,57 @@
...
-	 * @param moduleName
-	 *            value for native load pointing to the bootstrap code entry point
-	 * @return native code
-	 */
-	private String generateNativeLoad(String moduleName) {
-		if (Strings.isNullOrEmpty(moduleName))
+	private String generateBootCode(NodeRunOptions nodeRunOptions, Path workDir) throws IOException {
+		// 1 generate fake node project / folder
+		Path projectRootPath = workDir;
+		// 3 create 'node_modules' to the #1
```
[the remark this answers is not in the document]
Comment (yoosiba) [reply plugins/org.eclipse.n4js.runner.nodejs/src/org/eclipse/n4js/runner/nodejs/NodeEngineCommandBuilder.java]: ah yes... I had there more steps, but I have merged them.
Adjusted javadoc and comments.

File: plugins/org.eclipse.n4js.tester/src/org/eclipse/n4js/tester/TestDiscoveryHelper.java (hunk 2 above)
Comment (yoosiba) [reply plugins/org.eclipse.n4js.tester/src/org/eclipse/n4js/tester/TestDiscoveryHelper.java]: done

File: plugins/org.eclipse.n4js.transpiler.es/src/org/eclipse/n4js/transpiler/es/transform/ExpressionTransformation.xtend (hunk 4 above)
Comment (yoosiba) [reply plugins/org.eclipse.n4js.transpiler.es/src/org/eclipse/n4js/transpiler/es/transform/ExpressionTransformation.xtend]: done

File: plugins/org.eclipse.n4js.transpiler/src/org/eclipse/n4js/transpiler/AbstractTranspiler.java (hunk 3 above)
Comment (yoosiba) [reply plugins/org.eclipse.n4js.transpiler/src/org/eclipse/n4js/transpiler/AbstractTranspiler.java]: done

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/generator/AbstractSubGenerator.xtend [hunk 9]
```diff
@@ -37,39 +39,34 @@ import org.eclipse.xtext.validation.Issue
...
  */
 @Log
 abstract class AbstractSubGenerator implements ISubGenerator {
 
 	@Accessors
 	private CompilerDescriptor compilerDescriptor = null
 
-	@Inject
-	ProjectUtils projectUtils
+	@Inject protected ResourceNameComputer projectUtils
```
[the remark this answers is not in the document]
Comment (yoosiba) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/generator/AbstractSubGenerator.xtend]: removed (see next comment)

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/generator/AbstractSubGenerator.xtend [hunk 10]
```diff
@@ -37,39 +39,34 @@ import org.eclipse.xtext.validation.Issue
...
+	
+	@Inject protected StaticPolyfillHelper staticPolyfillHelper
 
-	@Inject
-	IN4JSCore core
+	@Inject protected IN4JSCore n4jsCore
 
-	@Inject
-	CompilerHelper compilerUtils
+	@Inject protected ResourceNameComputer resourceNameComputer
```
[the remark this answers is not in the document]
Comment (yoosiba) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/generator/AbstractSubGenerator.xtend]: adjusted

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/generator/AbstractSubGenerator.xtend [hunk 11]
```diff
@@ -223,23 +220,25 @@ abstract class AbstractSubGenerator implements ISubGenerator {
...
+		preferenceAccess.getPreference(input, getCompilerID(), CompilerProperties.COMPILED_FILE_SOURCEMAP_EXTENSION,
 			getDefaultDescriptor())
 	}
 
 	/** Adjust output-path of the generator to match the N4JS projects-settings. */
 	def void updateOutputPath(IFileSystemAccess fsa, String compilerID, Resource input) {
-		val outputPath = projectUtils.getOutputPath(input.URI)
+		var outputPath = n4jsCore.getOutputPath(input.URI)
+		if(outputPath === null)
+			outputPath = N4JSLanguageConstants.DEFAULT_PROJECT_OUTPUT
```
[the remark this answers is not in the document]
Comment (yoosiba) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/generator/AbstractSubGenerator.xtend]: done

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/naming/ModuleNameComputer.xtend [hunk 12]
```diff
@@ -29,6 +30,7 @@ import org.eclipse.xtext.resource.IResourceDescription
  * Client code should usually use {@code TModule.getName()} or {@code Script.getModule().getName()} to access the
  * module's name.
  */
+ @Singleton
```
[the remark this answers is not in the document]
Comment (yoosiba) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/naming/ModuleNameComputer.xtend]: done

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/projectModel/FindArtifactHelper.java (hunk 6 above)
Comment (yoosiba) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/projectModel/FindArtifactHelper.java]: done

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/utils/CompilerHelper.xtend [hunk 13]
```diff
@@ -1,101 +0,0 @@
...
-import org.eclipse.n4js.generator.ExceptionHandler
-import org.eclipse.n4js.projectModel.IN4JSProject
-import org.eclipse.n4js.projectModel.ProjectUtils
-import org.eclipse.n4js.resource.N4JSResource
-import org.eclipse.n4js.ts.types.TModule
-
-/**
- */
- @Singleton
-public class CompilerHelper {
```
[the remark this answers is not in the document]
Comment (yoosiba) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/utils/CompilerHelper.xtend]: gone. After refactoring made around `ResourceNameComputer` this was not used.

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/utils/ContainerTypesHelper.java [hunk 14]
```diff
@@ -91,7 +91,7 @@
 	/**
 	 * Injected to gain access for users of this class.
 	 */
-	public @Inject ProjectUtils projectUtils;
+	public @Inject ResourceNameComputer projectUtils;
```
[the remark this answers is not in the document]
Comment (yoosiba) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/utils/ContainerTypesHelper.java]: unused, deleted

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/validators/N4JSInjectorCallsitesValidator.xtend [hunk 15]
```diff
@@ -47,7 +47,7 @@ class N4JSInjectorCallsitesValidator extends AbstractN4JSDeclarativeValidator {
 	private N4JSTypeSystem ts;
 
 	@Inject
-	QualifiedNameComputer qualifiedNameComputer;
+	ResourceNameComputer qualifiedNameComputer;
```
[the remark this answers is not in the document]
Comment (yoosiba) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/validators/N4JSInjectorCallsitesValidator.xtend]: done

File: tests/org.eclipse.n4js.lang.tests/src/org/eclipse/n4js/naming/QualifiedNameComputerTest.java (hunk 5 above)
Comment (yoosiba) [reply tests/org.eclipse.n4js.lang.tests/src/org/eclipse/n4js/naming/QualifiedNameComputerTest.java]: I did rename the field, but I think is ok, especially that it tests the conversion to JS identifiers (the Unicode representation of problematic characters)

File: tests/org.eclipse.n4js.spec.examples.xpect.tests/testdata/my/Ex65_Object_Literal.n4js.xt (hunk 1 above)
Comment (yoosiba) [reply tests/org.eclipse.n4js.spec.examples.xpect.tests/testdata/my/Ex65_Object_Literal.n4js.xt]: in spec we have only parts of the code from tests. Parts I have changed are not there.

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/internal/AbstractN4JSCore.java [hunk 16]
```diff
@@ -137,9 +137,6 @@ public String getOutputPath(URI nestedLocation) {
 		if (container.isPresent()) {
 			outputPath = container.get().getProject().getOutputPath();
 		}
-		if (outputPath == null) {
-			outputPath = "src-gen";
-		}
```
[the remark this answers is not in the document]
Comment (yoosiba) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/internal/AbstractN4JSCore.java]: Errors on resources, right.
Compiling into folder that is not defined by the project is not desired IMHO, but I agree with avoiding errors on resources. 
I have added check in the `AbstractSubGenerator` to not generate compiled code if there is no output folder defined for the project. This way we avoid exceptions and generating files for misconfigured projects. (User still gets validation error in the manifest).

[the remark this answers is not in the document]
Comment (mor-n4) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/internal/AbstractN4JSCore.java]: ok!

File: tests/org.eclipse.n4js.lang.tests/src/org/eclipse/n4js/naming/QualifiedNameComputerTest.java (hunk 5 above)
Comment (mor-n4) [reply tests/org.eclipse.n4js.lang.tests/src/org/eclipse/n4js/naming/QualifiedNameComputerTest.java]: I meant: class should be renamed to ResourceNameComputerTest.

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/projectModel/ProjectUtils.xtend [hunk 17]
```diff
@@ -1,469 +0,0 @@
...
-import org.eclipse.xtext.resource.IEObjectDescription
-import org.eclipse.xtext.resource.impl.ResourceDescriptionsProvider
-
-import static extension org.eclipse.n4js.utils.N4JSLanguageUtils.*
-
-/**
- * Provides utilities for generating resource (file or module) descriptors that are based on containing project and its properties.
- */
-@Singleton
-public class ProjectUtils {
```
[the remark this answers is not in the document]
Comment (yoosiba) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/projectModel/ProjectUtils.xtend]: moved

**Pull request merged**